docs/architecture/0e-plans.md

```diff
@@ -568,6 +568,513 @@ Note that if we want to be able to remove all error code translations (to save c
+
+#### Changes to ASN.1 enablement
+
+Currently, `MBEDTLS_ASN1_PARSE_C` and `MBEDTLS_ASN1_WRITE_C` are automatically enabled:
+
+* If `MBEDTLS_RSA_C` is enabled. This is needed for RSA key import and export. We don't try to optimize code size when import or export of a key type is disabled, so this can stay. As part of internal refactoring, it will change to `PSA_WANT_KEY_TYPE_RSA_PUBLIC_KEY` eventually.
+* If `PSA_WANT_ALG_ECDSA` or `PSA_WANT_ALG_DETERMINISTIC_ECDSA` is built in. This used to be required to convert between ASN.1 and raw signature formats internally back when PSA was built on top of the legacy API, but today I think it isn't needed. However, it is needed for `MBEDTLS_ECDSA_C`. We do expose the signature conversion functions `mbedtls_ecdsa_raw_to_der` and `mbedtls_ecdsa_der_to_raw` when an ECDSA signature is enabled through PSA (built-in or not), but we automatically enable the necessary ASN.1 functions regardless of the ASN.1 configuration.
+
+So we have an optimization here to not auto-enable all of ASN.1 for built-in ECDSA. This will require removing or changing the guards on `mbedtls_ecdsa_read_signature()` and `mbedtls_ecdsa_write_signature()` and friends, which are now private functions.
+
+### Non-critical changes to compilation options
+
+The changes discussed in this section would make for a more user-friendly API, but would not reduce our maintenance burden. As a consequence, we are treating them as low priority. They may not be fully fleshed out.
+
+#### Revise client/server inclusion
+
+Now that PSA is “always on”, what does it mean for `MBEDTLS_PSA_CRYPTO_C`, `MBEDTLS_PSA_CRYPTO_CLIENT` and `MBEDTLS_PSA_CRYPTO_SPM`? We no longer officially support builds without any PSA parts, but we do support server-only or client-only builds, so there is no part of the code that is actually mandatory.
+
+* A library build (the most common case) defines `MBEDTLS_PSA_CRYPTO_C` (which implies `MBEDTLS_PSA_CRYPTO_CLIENT`).
+* A client-only build defines `MBEDTLS_PSA_CRYPTO_CLIENT` and little to no else.
+* A server-only build defines `MBEDTLS_PSA_CRYPTO_C` and `MBEDTLS_PSA_CRYPTO_SPM`.
+
+#### Revise key store selection
+
+For historical reasons, we have three key store architectures:
+
+* Purely static, when `MBEDTLS_PSA_STATIC_KEY_SLOTS` is enabled. Key slots are allocated statically, and the core does not use `malloc()`. This only allows a small number of keys determined at compile time, and keys have a size limit that is determined at compile time. This achieves the smallest code size, but has major limitations.
+* Hybrid, when neither `MBEDTLS_PSA_KEY_STORE_DYNAMIC` nor `MBEDTLS_PSA_STATIC_KEY_SLOTS` is enabled. This is the original design, with a static array of key slots (for simplicity) and dynamic allocation of key data. This only allows a small number of keys determined at compile time, but achieves a small code size (almost on par with the static key store if `malloc()` is present anyway).
```

Comment:
Does the hybrid architecture have any advantages over purely static? For example, maybe it has smaller runtime memory usage while retaining a smaller code size.

If it doesn't then perhaps it would make sense to remove this mode, as it's a bit of a historical accident.

Reply:
Expanded. All three have plausible scenarios where they're better than the other two.